Carrot User Documentation Feature (#18)
Health-Informatics-UoN/carrot: website/src/pages/carrot_etl_guide.mdx

> {/* ----- Imports ----- */}

Reviewer: I think you should include this page into the _meta.js file. It will help the page name on the sidebar menu nicer

Author: Okay then I'll add them in the _meta.js. Thanks for your feedback.

    - Understand what the security arrangements are.

    - Understand access rights to the dataset.

    - Define communication and feedback mechanisms and frequency thereof – these can be emails, weekly meetings, etc.
    </div>

</div>

  <div style={{ flex: "0.5", textAlign: "right" }}>
    <figure style={{ display: "inline-block", textAlign: "center" }}>
      <img 
        src="/docs/Carrot-SOP/3.png" 
        alt="Selecting Data during Data Pre-processing" 
        width="200"
        height="250"
        style={{ 
          maxWidth: "100%", 
          height: "auto", 
          border: "1px solid #ddd", 
          padding: "5px" 
        }} 
      />
      <figcaption style={{ marginTop: "8px", marginRight: "50px", fontStyle: "italic", color: "#666" }}>
        Figure 3: Selecting 
        Data during Data Pre-processing
      </figcaption>
    </figure>
  </div>

</div>

### Data Anonymisation

<div style={{ display: "flex", alignItems: "flex-start", justifyContent: "space-between", gap: "10px" }}>
  <div style={{ flex: "1" }}>

  The anonymisation of data is paramount, it is critical that the following steps are carefully followed.

>   <WarningIcon />It is the responsibility of the Data Owner to de-identify datasets before they are supplied for OMOP conversion.

Reviewer: I think we should use Callout component for the warning/any blocks that need attention. This will help to keep the consistent in the docs and reducing the need of having additional component, e.g., WarningIcon, favicon-warning-icon

For example, line 33-36, in the data standard page:
```
import { Callout } from "nextra/components";

<Callout type="warning">
  Ensure all Scan Report Files are thoroughly reviewed for any identifying
  information before extracting them from their secure location.
</Callout>
```

Author: Sure, I'll remove the component and replace with callout

      <img 
        src="/docs/Carrot-SOP/20.png" 
        alt="Check mapping rules in metadata mapping" 
        width="200"
        height="250"
        style={{ 
          maxWidth: "100%", 
          height: "auto", 
          border: "1px solid #ddd", 
          padding: "5px" 
        }} 
      />
      <figcaption style={{ marginTop: "8px", marginRight: "70px", fontStyle: "italic", color: "#666" }}>
        Figure 21: Check Mapping Rules
        in Metadata Mapping.
      </figcaption>
    </figure>
  </div>

</div>

<div style={{ display: "flex", flexDirection: "column", alignItems: "center", marginTop: "10px", marginRight: "40px" }}>
  <figure style={{ textAlign: "center", margin: "0 auto" }}>
    <img 
      src="/docs/Carrot-SOP/21.png" 
      alt="Example of Dataset Mapping Rule options" 
      width="700" 
      style={{ display: "block", margin: "0 auto" }} 
    />
    <figcaption style={{ marginTop: "8px", fontStyle: "italic", color: "#666" }}>
      Figure 22: Example of Dataset Mapping Rule Options.
    </figcaption>
  </figure>
</div>

{/* ----- End of Section ----- */}

{/* ----- Reference Section of the Documentation ----- */}

## Reference

> - Motwani, R., & Xu, Y. (2007). Efficient Algorithms for Masking and Finding Quasi-Iden.fiers. Available at: https://theory.stanford.edu/~xuying/papers/quasi_vldb1.pdf

Reviewer: Quasi-Identifiers?

Author: Oops! There was a parsing error in my documentation. I'll fix it now.

{/* ----- Imports ----- */}

import WarningIcon from '@/components/icons/WarningIcon';
import HighlightedText, { GreenHighlightedText } from '@/components/HighlightedText';

{/* ----- Landing View of the Documentation ----- */}

# Welcome to the Carrot’s Data Engineering Procedure for OMOP representation.

The Carrot’s Data Engineering Process for OMOP standard vocabulary process primarily consist of Data Pre-processing, Metadata Profiling and Metadata Mapping. 

<div>

> The following document sets the Standardised Opera.ng Procedure (SOP) that must be followed and applied during the engineering of data for conversion to an [OMOP](https://www.ohdsi.org/data-standardization/) representation. 

Reviewer 2: ```suggestion
The following document sets the Standardised Operating Procedure (SOP) that must be followed and applied during the engineering of data for conversion to an [OMOP](https://www.ohdsi.org/data-standardization/) representation. 
```

</div>

> Figure 1 illustrates the 3 main data engineering stages and the associated soQware tools that are used, this process is commonly known as an [Extract, Transform and Load (ETL) process](https://ohdsi.github.io/TheBookOfOhdsi/ExtractTransformLoad.html). 

Reviewer 2: ```suggestion
Figure 1 illustrates the 3 main data engineering stages and the associated software tools that are used, this process is commonly known as an [Extract, Transform and Load (ETL) process](https://ohdsi.github.io/TheBookOfOhdsi/ExtractTransformLoad.html). 
```

Author: 👍🏻

<div style={{ display: "inline-flex",  marginTop: "8px" }}>
  <WarningIcon />
  <span>This document only covers the ‘Extract and Transform’ sections of an ETL process as per the Figure below.</span>
</div>

<figure style={{ textAlign: "center" }}>
  <img 
    src="/docs/Carrot-SOP/omop-etl-process.png" 
    alt="ETL Process Diagram" 
    width="600" 
    style={{ display: "block", margin: "15px auto" }} 
  />
  <figcaption style={{ marginTop: "8px", fontStyle: "italic", color: "#666" }}>

>     Figure 1: The stages of Health Infomatics Data Engineering.

Reviewer 2: ```suggestion
    Figure 1: The stages of Health Informatics Data Engineering.
```

Author: 👍🏻

  </figcaption>
</figure>

<span style={{ paddingTop: "20px", display: "inline-block" }}>

>   <GreenHighlightedText> Moving to the Data Engineering Overview section -> </GreenHighlightedText>

Reviewer 2: What does this mean? I don't understand it in this context

Reviewer 2: Ah I see - these are like transitions before the next section? Could we just remove them, they don't seem to add much.

Author: Sure Andy. There are 3 transition texts at the moment and I'll remove them and push the code for PR.

</span>

{/* ----- End of Section ----- */}

{/* ----- Data Engineering Overview of the Documentation ----- */}

## Data Engineering Overview
Figure 2 portrays the high-level view of the 3 stages and the main processes they are comprised of:

<div style={{ display: "flex", alignItems:  "center", gap: "5px"}}>
<div style={{ flex: "1" }}>

- **Stage 1: Data Pre-processing**
<div>

> The pre-processing stage focuses on the ‘important aXributes’ of a dataset that need to be mapped to the OMOP Common Data Model for discoverability.

Reviewer 2: ```suggestion
The pre-processing stage focuses on the ‘important attributes’ of a dataset that need to be mapped to the OMOP Common Data Model for discoverability.
```

Author: 👍🏻

A meeting with the dataset owner is necessary to understand the context and select the key data attributes to be made discoverable. From this a subset of the data is then anonymised by the data owner and refactored (reorganised), producing a pre-processed dataset. This is then reviewed by the data owner to determine if it is fit for purpose and is fully anonymised.

The output of Stage 1 is an anonymised dataset ready for profiling together with a formaXed data dictionary.
</div>

- **Stage 2: Metadata Profiling**

<div>
The metadata profiling stage assesses the structure of a pre-processed dataset.

> The pre-processed dataset is loaded into the OHDSI White Rabbit tool, this produces a sta.s.cal profile of the data called a ‘Scan Report’ that further de-identifies data by removing the rela.onships between data attributes. This is then checked to ensure that the data has been rigorously anonymised.

Reviewer 2: ```suggestion
The pre-processed dataset is loaded into the OHDSI White Rabbit tool, this produces a statistical profile of the data called a ‘Scan Report’ that further de-identifies data by removing the relationships between data attributes. This is then checked to ensure that the data has been rigorously anonymised.
```

Author: 👍🏻

    1. **Meet with Data Owner**
    <div style={{ marginTop: "0px" }}>
      - Discuss their dataset(s) and specific requirements.<br/>
      - Understand what each dataset represents (so you understand their context and perspectives).<br/>
      - Explain the data engineering process (so they understand your perspective and what Health Informatics can offer).
    </div>

    2. **Choose Data Set(s)**
    <div style={{ marginTop: "0px" }}>
      - Discuss with the data owner which dataset(s) they would like converted.
    </div>
  

    3. **Select data and conversion goals:**
    <div style={{ marginTop: "0px" }}>
      - The aim is ‘Data Discoverability’.
      - Define conversion goals, i.e., what is to be converted for discoverability.
      - Select data attributes to be converted – not all data attributes are necessary for discoverability, a minimal set of data attributes may just be necessary.
    </div>

    4. **Define Working processes**
    <div style={{ marginTop: "0px" }}>
    - Understand where each dataset is located.

    - Understand what the security arrangements are.

    - Understand access rights to the dataset.

>     - Define communica.on and feedback mechanisms and frequency thereof – these can be emails, weekly meetings, etc.

Reviewer 2: ```suggestion
    - Define communication and feedback mechanisms and frequency thereof – these can be emails, weekly meetings, etc.
```

Author: 👍🏻

    </figcaption>
  </figure>
</div>

</div>

<div id="pii-section">
  ### Personally Identifiable Information (PII) / Personal Data

  This means any informaton relating to an identified or identifiable natural person (data subject); an identifiable natural person is one who can be identified, **directly or indirectly**, in particular by reference to an identifier such as a name, an identification number, location data, an online identifier or to one or more factors specific to the physical, physiological, genetic, mental, economic, cultural or social identity of that natural person [Information Commissioner's Office](https://ico.org.uk/for-organisations/uk-gdpr-guidance-and-resources/personal-information-what-is-it/what-is-personal-data/what-is-personal-data/).

  <figure style={{ textAlign: "center", margin: "15px auto" }}>
    <img 
      src="/docs/Carrot-SOP/5.png" 
      alt="Example of Direct and Indirect Identifiers" 
      width="500" 
      style={{ display: "block", margin: "0 auto" }} 
    />
    <figcaption style={{ marginTop: "8px", fontStyle: "italic", color: "#666" }}>
      Figure 5: Example of Direct and Indirect Identifiers
    </figcaption>
  </figure>

>   PII refers to data that can be used to identify, locate, or contact individuals or establishments, or reveal the characteristics or other details about them. PII might consist of direct iden.fiers, such as the name, social security number or other information that is unique to an individual. Indirect identifiers include uncommon race, ethnicity, extreme age, unusual occupa.on and other details [CDC](https://www.cdc.gov/nchs/training/confidentiality/training/page581.html).

Reviewer 2: ```suggestion
  PII refers to data that can be used to identify, locate, or contact individuals or establishments, or reveal the characteristics or other details about them. PII might consist of direct identifiers, such as the name, social security number or other information that is unique to an individual. Indirect identifiers include uncommon race, ethnicity, extreme age, unusual occupation and other details [CDC](https://www.cdc.gov/nchs/training/confidentiality/training/page581.html).
```

Author: 👍🏻

  <figure style={{ textAlign: "center", margin: "15px auto" }}>
    <img 
      src="/docs/Carrot-SOP/Indirect-Identifier.png" 
      alt="Example of Removing Indirect Identifiers" 
      width="400" 
      style={{ display: "block", margin: "0 auto" }} 
    />
    <figcaption style={{ marginTop: "8px", fontStyle: "italic", color: "#666" }}>
      Figure 7: Example of Removing Indirect Identifiers.
    </figcaption>
  </figure>

</div>

### Refactor Data

<div style={{ display: "flex", alignItems: "flex-start", justifyContent: "space-between", gap: "10px" }}>
  <div style={{ flex: "1" }}>

  <div style={{ marginTop: "0px" }}>
  The anonymisation of data is paramount, it is critical that the following steps are carefully followed.
  </div>
  <div style={{ marginTop: "0px" }}>
  <WarningIcon /> It is the responsibility of the Data Owner to de-identify datasets before they are supplied for OMOP conversion.
  </div>

  1. **Organise Data Tables**
  <div style={{ marginTop: "0px" }}>
  - Create a demographics table, consisting of:
      - Identifier
      - Date of Birth 
      - Sex at Birth 
      - Ethnicity
  <div>

>   - The rest of the data can then be arranged as deemed fit by the data owner and / or the CARROT data team, each table must include:

Reviewer 2: ```suggestion
  - The rest of the data can then be arranged as deemed fit by the data owner and / or the data team, each table must include:
```

Author: 👍🏻

      - 1 x identifying column per table.
      - 1 x date event per table.
  </div>
  </div>

  2. **Create Dataset Extract**
  <div style={{ marginTop: "0px" }}>

>   - First, check the [CARROT4OMOP Data Standards](https://carrot4omop.ac.uk/standards/#data-dictionary-file) page for exact requirements.

Reviewer 2: this should link internally to the standards page: 

https://carrot.ac.uk/standards

Author: Just replace the link instead of the previous link yeah?

Reviewer 2: Yep absolutely

Author: Alright👍🏻

      <img 
        src="/docs/Carrot-SOP/6.png" 
        alt="Refactoring the data in data preprocessing section" 
        width="200"
        height="250"
        style={{ 
          maxWidth: "100%", 
          height: "auto", 
          border: "1px solid #ddd", 
          padding: "5px" 
        }} 
      />
      <figcaption style={{ marginTop: "8px", marginRight: "70px", fontStyle: "italic", color: "#666" }}>
        Figure 8: Refactoring the 
        Data in Data Preprocessing.
      </figcaption>
    </figure>
  </div>

</div>

### Review

<div style={{ display: "flex", alignItems: "flex-start", justifyContent: "space-between", gap: "10px" }}>
  <div style={{ flex: "1" }}>

  <div>
  It is essential that all data extracts are reviewed prior to Stage 2.
  </div>

  <div style={{ marginTop: "0px" }}>
  1. **Review Dataset Extract**
  <div style={{ marginTop: "0px" }}>
  - Check the agreed data attributes chosen in Stage 1 are selected.
  - Check that the dataset has been anonymised as per Stage 2.
  - Check the demographics table exists and contains these 4 data attributes:

      - Identifier
      - Date of Birth
      - Sex
      - Ethnicity

>   - Check that there is a date event to map to per dataset table and that it conforms to [CARROT4OMOP](https://carrot4omop.ac.uk/standards/#data-dictionary-file).

Reviewer 2: This link should also be: https://carrot.ac.uk/standards

Author: 👍🏻

        Figure 13: Checking Scan 
        Report in Metadata Profiling.
      </figcaption>
    </figure>
  </div>

</div>

<div style={{ display: "flex", flexDirection: "column", alignItems: "center", marginTop: "10px", marginRight: "10px" }}>
  <figure style={{ textAlign: "center", margin: "0 auto" }}>
    <img 
      src="/docs/Carrot-SOP/13.png" 
      alt="Example of WhiteRabbit Scan Report - Aspects to Check" 
      width="700" 
      style={{ display: "block", margin: "0 auto" }} 
    />
    <figcaption style={{ marginTop: "8px", fontStyle: "italic", color: "#666" }}>
      Figure 14: Example of WhiteRabbit Scan Report - Aspects to Check.
    </figcaption>
  </figure>
</div>

<span style={{ paddingTop: "20px", display: "inline-block" }}>
  <GreenHighlightedText> Moving to the Metadata Mapping Section -> </GreenHighlightedText>
</span>

{/* ----- End of Section ----- */}

{/* ----- Metadata Section of the Documentation ----- */}

## Stage 3: Metadata Mapping

<div style={{ display: "flex", alignItems: "flex-start", justifyContent: "space-between", gap: "20px" }}>
  <div style={{ flex: "1" }}>

  ### Upload to Carrot Scan Report

  **Upload Scan Report**

>   - Log into [CARROT](https://ccom.azurewebsites.net/):

Reviewer 2: ```suggestion
  - Log into [Carrot](https://carrot.ac.uk):
```

Author: Ok

  - Upload Scan Report:
    - Click **Scan Reports** Tab.
    - Click **New Scan Report Upload** option.
    - Choose **Data Partner** - from drop-down list. 
    - Choose **Dataset** - from drop-down list.
    - Choose **Visibility**: Public or Restricted

>       - Public = All registered users of CARROT can view the scan report.

Reviewer 2: ```suggestion
      - Public = All registered users of Carrot can view the scan report.
```

Carrot always lowercase

Author: Ok Andy👍🏻

        width="200"
        height="250"
        style={{ 
          maxWidth: "100%", 
          height: "auto", 
          border: "1px solid #ddd", 
          padding: "5px" 
        }} 
      />
      <figcaption style={{ marginTop: "8px", marginRight: "60px", fontStyle: "italic", color: "#666" }}>
        Figure 15: Upload Mapping
        in Metadata Mapping.
      </figcaption>
    </figure>
  </div>

</div>

<div style={{ display: "flex", flexDirection: "column", alignItems: "center", marginTop: "10px", marginRight: "10px" }}>
  <figure style={{ textAlign: "center", margin: "0 auto" }}>
    <img 
      src="/docs/Carrot-SOP/15.png" 
      alt="Example of CARROT software tool screens for New Scan Report Upload" 
      width="700" 
      style={{ display: "block", margin: "0 auto" }} 
    />
    <figcaption style={{ marginTop: "8px", fontStyle: "italic", color: "#666" }}>
      Figure 16: Example of CARROT Software Tool Screens for New Scan Report Upload.
    </figcaption>
  </figure>
</div>

<div style={{ display: "flex", alignItems: "flex-start", justifyContent: "space-between", gap: "20px" }}>
  <div style={{ flex: "1" }}>

  ### Check the Scan Report Upload

>   While logged into the CARROT tool:

Reviewer 2: ```suggestion
  While logged into the Carrot tool:
```

Author: 👍🏻

  - Click **Scan Reports**.
  - Select **Scan Reports** tab.
  - Check the progress of your Scan Report upload – it will show one of the following three (see diagram): 
    - <span style={{ color: "green", fontWeight: "bold" }}>Upload Complete</span>
    - <span style={{ color: "orange", fontWeight: "bold" }}>Upload In Progress</span>
    - <span style={{ color: "red", fontWeight: "bold" }}>Upload Failed</span>
  - Refresh the web browser to check on upload progress.
  </div>

  <div style={{ flex: "0.5", textAlign: "right", alignSelf: "flex-start", marginTop: "40px" }}>
    <figure style={{ display: "inline-block", textAlign: "center", margin: "0 auto" }}>
      <img 
        src="/docs/Carrot-SOP/16.png" 
        alt="Upload to CARROT in metadata mapping" 
        width="160"
        height="190"
        style={{ 
          maxWidth: "100%", 
          height: "auto", 
          border: "1px solid #ddd", 
          padding: "5px" 
        }} 
      />
      <figcaption style={{ marginTop: "8px",marginRight: "100px", fontStyle: "italic", color: "#666" }}>

>         Figure 17: Upload to CARROT

Reviewer 2: ```suggestion
        Figure 17: Upload to Carrot
```

Author: 👍🏻

        in Metadata Mapping.
      </figcaption>
    </figure>
  </div>

</div>

<div style={{ display: "flex", flexDirection: "column", alignItems: "center", marginTop: "20px", marginRight: "40px" }}>
  <figure style={{ textAlign: "center", margin: "0 auto" }}>
    <img 
      src="/docs/Carrot-SOP/17.png" 
      alt="Example of CARROT tool Scan Report Status page" 
      width="700" 
      style={{ display: "block", margin: "0 auto" }} 
    />
    <figcaption style={{ marginTop: "8px", fontStyle: "italic", color: "#666" }}>

>       Figure 18: Example of CARROT Tool Scan Report Status Page

Reviewer 2: ```suggestion
      Figure 18: Example of Carrot Tool Scan Report Status Page
```

Author: 👍🏻

    </figcaption>
  </figure>
</div>

<div style={{ display: "flex", alignItems: "flex-start", justifyContent: "space-between", gap: "20px" }}>
  <div style={{ flex: "1" }}>

    ### Map Rules

    #### Choose Dataset

>     - Log into [CARROT](https://ccom.azurewebsites.net/):

Reviewer 2: ```suggestion
    - Log into [Carrot](https://carrot.ac.uk):
```

Author: 👍🏻

        height="190"
        style={{ 
          maxWidth: "100%", 
          height: "auto", 
          border: "1px solid #ddd", 
          padding: "5px" 
        }} 
      />
      <figcaption style={{ marginTop: "8px", marginRight: "100px", fontStyle: "italic", color: "#666" }}>
        Figure 19: Upload Mapping Rules in
        Metadata Mapping.
      </figcaption>
    </figure>
  </div>

</div>

<div style={{ display: "flex", flexDirection: "column", alignItems: "center", marginTop: "10px", marginRight: "40px" }}>
  <figure style={{ textAlign: "center", margin: "0 auto" }}>
    <img 
      src="/docs/Carrot-SOP/19.png" 
      alt="Setting Person ID and Date Event for each Dataset Metadata Table" 
      width="700" 
      style={{ display: "block", margin: "0 auto" }} 
    />
    <figcaption style={{ marginTop: "8px", fontStyle: "italic", color: "#666" }}>
      Figure 20: Setting Person ID and Date Event for Each Dataset Metadata Table.
    </figcaption>
  </figure>
</div>

<div style={{ display: "flex", alignItems: "flex-start", justifyContent: "space-between", gap: "20px" }}>
  <div style={{ flex: "1" }}>

    ### Check Map Rules

    #### Choose Dataset

>     - Log into [CARROT]():

Reviewer 2: ```suggestion
    - Log into [Carrot](https://carrot.ac.uk):
```

Author: 👍🏻
